core/src/consensus.rs

```diff
@@ -642,6 +642,22 @@ impl Tower {
                     SwitchForkDecision::FailedSwitchDuplicateRollback(latest_duplicate_ancestor)
                 };
 
+                // `heaviest_subtree_fork_choice` entries are not cleaned by duplicate block purging/rollback logic,
+                // so this is safe to check here. We return here if the last voted slot was rolled back/purged due to
+                // being a duplicate because `ancestors`/`descendants`/`progress` structurs may be missing this slot due
+                // to duplicate purging. This would cause many of the `unwrap()` checks below to fail.
+                if let Some(latest_duplicate_ancestor) = heaviest_subtree_fork_choice.latest_invalid_ancestor(&(last_voted_slot, last_voted_hash)) {
+                    // We're rolling back because one of the ancestors of the last vote was a duplicate. In this
+                    // case, it's acceptable if the switch candidate is one of ancestors of the previous vote,
+                    // just fail the switch check because there's no point in voting on an ancestor. ReplayStage
+                    // should then have a special case continue building an alternate fork from this ancestor, NOT
+                    // the `last_voted_slot`. This is in contrast to usual SwitchFailure where ReplayStage continues to build blocks
+                    // on latest vote. See `select_vote_and_reset_forks()` for more details.
+                    if heaviest_subtree_fork_choice.is_ancestor_slot(switch_slot, &(last_voted_slot, last_voted_hash)) {
```

Comment:
Yeah this is a subtle point.

It is possible for ReplayStage to call `check_switch_threshold()` on a smaller slot than your last voted slot if it's trying to switch to a different fork. In these cases the switch threshold succeeds, but then the lockout check fails. The subtle difference here is that even though the lockout check fails in both cases, if the switch threshold succeeds we build our next block on the other fork, if it fails, we continue building blocks on our fork. This may not be a critical distinction now that we ingest votes for switching threshold from gossip (before you needed somebody to build blocks to see switching votes), but hesitant to make that change here.

So for this change:

What should never happen is you try to switch to smaller slot than your last voted slot AND that smaller slot is an ancestor of your last vote. This correctness check is enforced by this panic https://github.com/solana-labs/solana/blob/b2d071a9ea01c622b90cbece49f313b5bb886afb/core/src/consensus.rs#L690-L695. The only exception to this rule is the rollback based on dupe slots, which is why there's a special callout for that case and *only* this case here.

Reply:
Got it, thanks for pointing at that panic, sounds good